src/node.cpp

```diff
@@ -1554,7 +1554,8 @@ LocalNode::~LocalNode()
         return;
     }
 
-    if (sync->state == SYNC_ACTIVE || sync->state == SYNC_INITIALSCAN)
+    if (!sync->mDestructorRunning && (
```

Comment:
I agree, there are a high proportion of lines of code in that function which are calls into the sync.  However, I think that is a task for another refactoring PR rather than this one, which has already accumulated many changes (and many more to come with the combining of SyncConfig and JSONSyncConfig).